support loading of stan playground runs

magland opened the pull request:
For example:

http://localhost:5173/mcmc-monitor?s=spa#/spa/vzundlxu/main.spa.out
when merged and deployed this will become
https://flatironinstitute.github.io/mcmc-monitor?s=spa#/spa/vzundlxu/main.spa.out

Here `s=spa` means that we are not connecting to a service, but rather we are going to retrieve the data from the stan playground database (spa stands for stan playground analysis)

The `/spa/vzundlxu/main.spa.out` path is referring to a particular project ID and output file name.

I tried to interfere as little as possible with the existing flow. The main new stuff is in the src/spaInterface folder. That has logic to query the serverless API of stan playground (on vercel).

Internally, the runId takes the form `spa|{projectId}|{fileName}` and then in a couple of places we detect that format and then retrieve data using the alternative mechanism. The state management (dispatch, etc) is not changed.

magland committed:
support loading of stan playground runs

magland committed:
handle s=spa, empty service URL

jsoules commented on src/MCMCMonitorDataManager/updateSequences.ts:
```diff
@@ -1,25 +1,39 @@
-            sequences: X.map(s => ({
-                runId: s.runId, chainId: s.chainId, variableName: s.variableName, position: s.data.length
-            }))
+        const runId = X[0].runId
```
There's not actually anything enforcing that the `runId`s be homogeneous across all sequences--so technically it's insufficient to check the first one only (even though I have a hard time imagining that a situation where we have a mix would come about). However, I would recommend changing this to a `some`/`every` situation so that we can be explicit and catch it if the case does ever come up.

jsoules commented on src/MCMCMonitorDataManager/updateSequences.ts:
```diff
@@ -1,25 +1,39 @@
-                runId: s.runId, chainId: s.chainId, variableName: s.variableName, position: s.data.length
-            }))
+        const runId = X[0].runId
+        if (runId.startsWith('spa|')) {
```
This does appear in multiple places, so probably a good idea to turn it into a utility function. (We could use `SpaRun` as a subtype of `MCMCRun` and use a typeguard for it, but we won't actually have the full run object for this test anyway.)

jsoules reviewed (commented):
Mostly questions for discussion

magland commented on src/MCMCMonitorDataManager/updateSequences.ts:
```diff
@@ -1,25 +1,39 @@
+            // handle the special case of a stan playground run
+            const sequenceUpdates: MCMCSequenceUpdate[] | undefined = await getSpaSequenceUpdates(runId, X)
+            if (sequenceUpdates) {
+                dispatch({
+                    type: "updateSequenceData",
+                    sequences: sequenceUpdates
+                })
```
```
const sequenceUpdtates = isSpaRunId(runId) ? ... : ...
```

magland replied:
I am actually going to leave as an if branch for now, since the two branches are sufficiently different, but I will pull out the common dispatch call.

magland committed:
address js review of #85

magland commented:
@jsoules I pushed a commit addressing all of your points. The only one I didn't fully address is avoiding the branching code between spa and non-spa cases. I did however reduce the duplicated line of code to one line for those code blocks.

magland committed:
include spaMode in tests config

magland commented:
@jsoules not sure where the 'window is not defined' errors are coming from in the automated tests...

magland committed:
set test env to jsdom in vite.config.ts

jsoules commented on vite.config.ts:
```diff
@@ -20,7 +20,8 @@ export default defineConfig({
     },
-    exclude: [...configDefaults.exclude, "service/**"]
+    exclude: [...configDefaults.exclude, "service/**"],
+    environment: 'jsdom'
```
I'd prefer to keep the default environment as `node` and use the `jsdom` environment only for individual UI-oriented files, as with the [ChainsSelector.test.tsx](https://github.com/flatironinstitute/mcmc-monitor/blob/0f79b0b0b0f9a19e49ad5f85333f4c4f27062def/test/components/ChainsSelector.test.tsx#L1) example.

Also, depending on the circumstance, these issues can occur during the import chain for logic that we actually want to mock out anyway; see for instance the mocks of `config` and `fetch` in the [postApiRequest.test.ts](https://github.com/flatironinstitute/mcmc-monitor/blob/0f79b0b0b0f9a19e49ad5f85333f4c4f27062def/test/networking/postApiRequest.test.ts) tests.

If you want to call out specific things that were failing, I can take a look after my meeting and make some suggestions?

jsoules committed:
Revert to node.js default environment. Mock import to avoid error.

magland committed:
Merge pull request #86 from flatironinstitute/stan-playground-spa-js

Revert to node.js default environment. Mock import to avoid error.

jsoules reviewed (commented):
I think this looks good. Do you have a link for where I could see it in action?  Just want to check it out in the browser before I merge.

Thanks!

magland commented:
Here's a link to the dev folder:
https://flatironinstitute.github.io/mcmc-monitor/dev/?s=spa#/spa/ffzkztff/main.spa.out

If it looks good, then please merge and deploy.

jsoules commented:
Merged and published. Thanks!
